### PR TITLE
Add `RelativeAverageSpectralError` and `RootMeanSquaredErrorUsingSlidingWindow` 

### LINKED ISSUE #799: More image metrics
## 🚀 Feature

just found this package and seems we have not yet all Image metrics :] 
some inspiration came from https://github.com/andrewekhalel/sewar but we aim on own implementation with `torch`

### Alternatives

- ~Mean Squared Error (MSE)~
- ~Root Mean Sqaured Error (RMSE)~
- ~Peak Signal-to-Noise Ratio (PSNR)~
- ~Structural Similarity Index (SSIM)~
- ~Universal Quality Image Index (UQI)~
- ~Multi-scale Structural Similarity Index (MS-SSIM)~
- ~Erreur Relative Globale Adimensionnelle de Synthèse (ERGAS)~
- ~Spatial Correlation Coefficient (SCC)~
- ~Relative Average Spectral Error (RASE)~
- ~Spectral Angle Mapper (SAM)~
- ~Spectral Distortion Index (D_lambda)~
- ~Spatial Distortion Index (D_S)~
- ~Quality with No Reference (QNR)~
- ~Visual Information Fidelity (VIF)~
- ~Block Sensitive - Peak Signal-to-Noise Ratio (PSNR-B)~

### Additional context

any of these metrics would be a separate PR/addition :]

### PR DESCRIPTION
**Pull request by Piyush-97**
## What does this PR do?

This PR adds a new metric Relative Average Spectral Error (RASE), and Root Mean Squared Error using Sliding Window the metric RASE is based upon.

Resolves part of #799

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by SkafteNicki** [comment]
Hi @Piyush-97, thanks for wanting to contribute.
However, I cannot see that you have made any changes. Is that correct?

**Comment by Piyush-97** [comment]
Yes, Sorry about that I was busy with some other work and couldn't give mush time to this.
You can assign this to someone else if this need to be completed fast.
Sorry again.

**Comment by Piyush-97** [comment]
> Hi @Piyush-97, thanks for wanting to contribute. However, I cannot see that you have made any changes. Is that correct?

Yes, Sorry about that I was busy with some other work and couldn't give mush time to this.
You can assign this to someone else if this need to be completed fast.
Sorry again.

File: torchmetrics/image/rase.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        tensor(8091.1338)
+
+    Raises:
+        ValueError: If ``window_size`` is not a positive integer.
+    """
+
+    total_images: Tensor
+    rmse_map: Tensor = None
+    target_sum: Tensor = None
+    higher_is_better = False
```
Comment (SkafteNicki) [inline torchmetrics/image/rase.py:52]: it seem differentiable to me?
```suggestion
    higher_is_better: bool = False
    is_differentiable: bool = True
```

File: torchmetrics/image/rmse_sw.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        tensor(0.4008)
+
+    Raises:
+        ValueError: If ``window_size`` is not a positive integer.
+    """
+
+    rmse_val_sum: Tensor
+    rmse_map: Tensor = None
+    total_images: Tensor
+    higher_is_better = False
```
Comment (SkafteNicki) [inline torchmetrics/image/rmse_sw.py:53]: same here
```suggestion
    higher_is_better: bool = False
    is_differentiable: bool = True
```

**Review by SkafteNicki — APPROVED** [review]
Minor comments, else LGTM

File: torchmetrics/image/rmse_sw.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        """
+        if self.return_rmse_map and self.rmse_map is None:
+            _img_shape = target.shape[1:]  # channels, width, height
+            self.rmse_map = torch.zeros(_img_shape, dtype=target.dtype, device=target.device)
+
+        self.rmse_val_sum, self.rmse_map, self.total_images = _rmse_sw_update(
+            preds, target, self.window_size, self.rmse_val_sum, self.rmse_map, self.add_total_images
+        )
+
+    def _compute(self) -> Union[Tensor, Tuple[Tensor, Tensor]]:
```
Comment (SkafteNicki) [inline torchmetrics/image/rmse_sw.py:87]:
```suggestion
    def compute(self) -> Union[Tensor, Tuple[Tensor, Tensor]]:
```

File: torchmetrics/image/rmse_sw.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        self.add_state("rmse_val_sum", default=torch.tensor(0.0), dist_reduce_fx="sum")
+        self.add_state("add_total_images", default=torch.tensor(0.0), dist_reduce_fx="sum")
+
+        if not isinstance(window_size, int) or isinstance(window_size, int) and window_size < 1:
+            raise ValueError("Argument `window_size` is expected to be a positive integer.")
+
+        self.window_size = window_size
+        self.return_rmse_map = return_rmse_map
+
+    def _update(self, preds: Tensor, target: Tensor) -> None:  # type: ignore
```
Comment (SkafteNicki) [inline torchmetrics/image/rmse_sw.py:72]:
```suggestion
    def update(self, preds: Tensor, target: Tensor) -> None:  # type: ignore
```

File: torchmetrics/image/rase.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        if self.rmse_map is None:
+            _img_shape = target.shape[1:]  # channels, width, height
+            self.rmse_map = torch.zeros(_img_shape, dtype=target.dtype, device=target.device)
+            self.target_sum = torch.zeros(_img_shape, dtype=target.dtype, device=target.device)
+
+        self.rmse_map, self.target_sum, self.total_images = _rase_update(
+            preds, target, self.window_size, self.rmse_map, self.target_sum, self.add_total_images
+        )
+
+    def _compute(self) -> Tensor:
```
Comment (SkafteNicki) [inline torchmetrics/image/rase.py:86]:
```suggestion
    def compute(self) -> Tensor:
```

File: torchmetrics/image/rase.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        super().__init__(compute_on_step=compute_on_step, **kwargs)
+        self.add_state("add_total_images", default=torch.tensor(0.0), dist_reduce_fx="sum")
+
+        if not isinstance(window_size, int) or isinstance(window_size, int) and window_size < 1:
+            raise ValueError("Argument `window_size` is expected to be a positive integer.")
+
+        self.window_size = window_size
+        self.return_rmse_map = return_rmse_map
+
+    def _update(self, preds: Tensor, target: Tensor) -> None:  # type: ignore
```
Comment (SkafteNicki) [inline torchmetrics/image/rase.py:70]:
```suggestion
    def update(self, preds: Tensor, target: Tensor) -> None:  # type: ignore
```

**Comment by Borda** [comment]
@Piyush-97 is it still WIP or ready to go? :rabbit:

**Comment by stancld** [comment]
> @Piyush-97 is it still WIP or ready to go? 🐰

I've been working on this PR.. I'll mark the PR as ready for review and try to find out where the problem with failing tests is.

**Comment by Borda** [comment]
@Piyush-97, how is it going here? :rabbit:

**Comment by Borda** [comment]
@Piyush-97 how is it going here? :rabbit: 
we are considering a new feature release next week, do you think we can finish this addition by then so it will be included? :pray:

**Comment by Borda** [comment]
@Piyush-97 I think we are on a good path, mind checking the collisions and failing tests?

**Comment by Borda** [comment]
@Piyush-97, could you pls check and resolve conflicts? :otter:

**Review by japui-coder — COMMENTED** [review]
This merge commit was added into this branch cleanly. Approved.

**Comment by stancld** [comment]
Found out we have a bug in `uniform_filter`, which returns inconsistent results with a scipy implementation of this filter. Trying to figure out what's wrong there

**Comment by Borda** [comment]
> Found out we have a bug in `uniform_filter`, which returns inconsistent results with a scipy implementation of this filter. Trying to figure out what's wrong there

interesting, do you have a draft PR with debugging? :rabbit:

**Comment by stancld** [comment]
@Borda I'm not sure, how they compute `uniform_filter` for multi-channel images
image

**Comment by Borda** [comment]
> Already found a proper way how to compute that. I'll clean the code tomorrow and push it here :]

Awesome! :tada:

**Comment by Borda** [comment]
> Already found a proper way how to compute that. I'll clean the code tomorrow and push it here :]

@stancld so what is the way we can finish it? :otter:

**Review by Borda — APPROVED** [review]

**Pull request merged**